lang/py/src/avro/schema.py

```diff
@@ -514,18 +506,16 @@ def __eq__(self, that):
 
 class FixedDecimalSchema(FixedSchema, DecimalLogicalSchema):
   def __init__(self, size, name, precision, scale=0, namespace=None, names=None, other_props=None):
+    max_precision = round(floor(log10(pow(2, (8 * max(1, size) - 1)) - 1)))
```

Comment:
Can we change the import statements here to do `import math` and then `math.log10` and `math.floor`? That way it'll be easier to tell when, for example, we're using `__builtins__.pow` vs `math.pow`, which I had to look up just now :)

Reply:
That said, this math is interesting. I don't understand where the formula is really coming from, but applying the logarithm product rule we can eliminate the exponent entirely and write it as

```python
max_precision = math.floor(math.log10(2) * (8 * max(1, size) - 1)) - 1
```

Please let me know if I missed something.

Reply:
Nice math catch!  As a side benefit, we can get rid of the `max` if the `pow` is gone.

This formula looks like the literal translation of https://github.com/apache/avro/blob/a2098bd88361b5ff9298c462a2dccafbb0c2c508/lang/java/avro/src/main/java/org/apache/avro/LogicalTypes.java#L262 (and your optimisation can be applied there too!).

Reply:
I don’t understand why we can get rid of `max`. Or rather, shouldn’t we raise an exception if size is negative, rather than inferring anything about it in the math itself?

Reply:
We _should_ raise the same exception as before when size is invalid.

Before applying the log power rule, we could end up with a negative value in the `log10` function if size was invalid, resulting in a `ValueError: math domain error` instead of the expected `SchemaParseException` or `AvroException`.  The only purpose of the max was to let the logical type be marked as invalid to get the exception from the underlying type.

When trying the FixedSchema, it correctly raises the AvroException("Fixed Schema requires a valid positive integer for size property")

(But I tend to agree: it looks OK when swallowing SchemaParseExceptions, but it gives a nonsensical warning message when deciding the schema is invalid!)

Reply:
**In either case, I found a math error in the formula...**  :/

```
>>> size = 1
>>> round(floor(log10(pow(2, (8 * max(1, size) - 1)) - 1)))
2
>>> math.floor(math.log10(2) * (8 * size - 1)) - 1
1
```

As reference values, I'd expect max_precision to be 2 for 1 byte, 9 for 4 bytes, and 18 for 8 bytes (which are not given by the last commit!).

Reply:
Good catch